### PR TITLE
Remove id_window Decorator

### PR DESCRIPTION
**Pull request by ghickman**
This replaces the `id_window` decorator with some composable functions and aims to extract the core functionality from id_window making them reusable.

Along the way I also documented everything in `matchers.py`, using the `pep257` tool to see what it was like (pretty good!) and added another QuerySet chunking function that works in the opposite direction.

### CODE REVIEW

File: treeherder/autoclassify/matchers.py [hunk 1]
```diff
@@ -55,6 +56,34 @@ def query_best(self, text_log_error):
...
+    if matches is None:
+        return
+
+    match = matches.first()
+    if match is None:
+        return
+
+    # generate a new score from the current match
+    dividend, divisor = score_multiplier
+    score = match.score * dividend / divisor
```
Comment (edmorley) [inline treeherder/autoclassify/matchers.py]: I was about to say we should probably `from __future__ import division` - but I see it's already in this file :-)

File: treeherder/autoclassify/matchers.py [hunk 2]
```diff
@@ -52,118 +66,130 @@ def match(self, text_log_error):
...
 
-        return None
+    # list of (match, score) pairs
+    # get the best score
+    match_and_score = first(matches, key=lambda m: (-m[1], -m[0].classified_failure_id))
+    if not match_and_score:
+        return
 
+    match, score = match_and_score
+    # (classified_failure.id, score)
```
Comment (edmorley) [inline treeherder/autoclassify/matchers.py]: This comment seems redundant?

Comment (ghickman) [reply treeherder/autoclassify/matchers.py]: Oops, left over from testing!

Comment (ghickman) [reply treeherder/autoclassify/matchers.py]: Fixed and rebased.

**Pull request merged**